ext/opcache/zend_persist.c

```diff
@@ -923,7 +923,28 @@ zend_class_entry *zend_persist_class_entry(zend_class_entry *orig_ce)
 		ZEND_HASH_MAP_FOREACH_BUCKET(&ce->constants_table, p) {
 			ZEND_ASSERT(p->key != NULL);
 			zend_accel_store_interned_string(p->key);
-			zend_persist_class_constant(&p->val);
+
+			zend_class_constant *c = Z_PTR(p->val);
+			if (c->ce == orig_ce || (Z_CONSTANT_FLAGS(c->value) & CONST_OWNED)) {
+				zend_persist_class_constant(&p->val);
+			} else {
+				zend_class_constant *new_c = zend_shared_alloc_get_xlat_entry(c);
+				if (new_c) {
+					Z_PTR(p->val) = new_c;
+				} else if (c->ce->ce_flags & ZEND_ACC_IMMUTABLE) {
+					/* Existing persistent class constant, keep the pointer to the old one. */
+				} else {
+#ifndef ZEND_WIN32
+					/* Internal class constant, keep the pointer to the old one. */
+					ZEND_ASSERT(c->ce->type == ZEND_INTERNAL_CLASS);
+#else
+					/* Windows must not persist linked classes with internal parents. We must not
+					 * point to internal structures in shm on Windows, because they might live in
+					 * different places for each process. */
+					ZEND_UNREACHABLE();
+#endif
+				}
+			}
```

Comment:
why not to move all this logic into ``zend_persist_class_constant()`` and ``zend_persist_class_constant_calc()``?
It already checks for ``zend_shared_alloc_get_xlat_entry(), so you duplicate this logic. Thinking about this, I'm not sure that ``(c->ce == orig_ce || (Z_CONSTANT_FLAGS(c->value) & CONST_OWNED))`` is the right condition at all.
In case you like to restrict cloning of internal constants, as you write in RFC comments, it's clearly to check ``c->ce->type == ZEND_INTERNAL_CLASS``.